api/retry.go

```diff
@@ -0,0 +1,110 @@
+	// maxBetweenCalls sets a capping value for the delay between calls, to avoid it growing infinitely
+	maxBetweenCalls time.Duration // Optional, defaults to 0, meaning no time cap
+	// maxToLastCall sets a capping value for all the retry process, in case there is a deadline to make the call.
+	maxToLastCall time.Duration // Optional, defaults to 0, meaning no time cap
+	// fixedDelay is used in case an uniform distribution of the calls is preferred.
+	fixedDelay time.Duration // Optional, defaults to 0, meaning Delay is exponential, starting at 1sec
+	// delayBase is used to calculate the starting value at which the delay starts to grow,
+	// When left empty, a value of 1 sec will be used as base and then the delays will
+	// grow exponentially with every attempt: starting at 1s, then 2s, 4s, 8s...
+	delayBase time.Duration // Optional, defaults to 1sec
+}
+
+func (c *Client) retryPut(ctx context.Context, endpoint string, in, out any, q *WriteOptions) (*WriteMeta, error) {
+	var err error
+	var wm *WriteMeta
+
+	attemptDelay := time.Duration(0)
+	startTime := time.Now()
+
+	t := time.NewTimer(attemptDelay)
+	defer t.Stop()
+
+	for attempt := uint64(0); attempt < c.config.retryOptions.maxRetries; attempt++ {
+		attemptDelay = c.calculateDelay(attempt)
+
+		t.Reset(attemptDelay)
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-t.C:
+
+		}
+
+		wm, err = c.put(endpoint, in, out, q)
+
+		// Maximum retry period is up, don't retry
+		if c.config.retryOptions.maxToLastCall != 0 && time.Now().Sub(startTime) > c.config.retryOptions.maxToLastCall {
+			break
+		}
+
+		// The put function only returns WriteMetadata if the call was successful
+		// don't retry
+		if wm != nil {
+			break
+		}
+
+		// If WriteMetadata is nil, we need to process the error to decide if a retry is
+		// necessary or not
+		var callErr UnexpectedResponseError
+		ok := errors.As(err, &callErr)
+
+		// If is not UnexpectedResponseError, it is an error while performing the call
+		// don't retry
+		if !ok {
+			break
+		}
+
+		// Only 500+ or 429 status calls may be retried, otherwise
+		// don't retry
+		if !isCallRetriable(callErr.StatusCode()) {
+			break
+		}
+	}
+
+	return wm, err
+}
+
+// According to the HTTP protocol, it only makes sense to retry calls
+// when the error is caused by a temporary situation, like a server being down
+// (500s+) or the call being rate limited (429), this function checks if the
+// statusCode is between the errors worth retrying.
+func isCallRetriable(statusCode int) bool {
+	return statusCode > http.StatusInternalServerError &&
+		statusCode < http.StatusNetworkAuthenticationRequired ||
+		statusCode == http.StatusTooManyRequests
+}
+
+func (c *Client) calculateDelay(attempt uint64) time.Duration {
+	if c.config.retryOptions.fixedDelay != 0 {
+		return c.config.retryOptions.fixedDelay
+	}
+
+	if attempt == 0 {
+		return 0
+	}
+
+	newDelay := c.config.retryOptions.delayBase << (attempt - 1)
```

Comment:
I believe this will overflow quite quickly similarly to what is shown here: https://github.com/hashicorp/nomad/pull/18200

Reply:
Good catch. @Juanadelacuesta note that we don't want to pull in the `helper` into this `api` package, so we'll need to duplicate the logic we added in #18200 here.

Reply:
Should we set a limit to the max number of attempts?